From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Christopher Cameron <ccameron@chromium.org>
Date: Fri, 10 Apr 2026 08:22:48 +0000
Subject: skhdr::AdaptiveGlobalToneMap: Fix OOB write

In MakeGainCurveXYMImage we:
- compute the number of control points used
- allocate an image of the spec's maximum number of control points
- write data for the number of control points used

If a gain curve is longer than the spec's maximum number of control
points, then this will write out of bounds.

This will be fixed in two ways.

The first fix is to allocate the image size to be the size that will be
written.

The second fix is to surface a AdaptiveGlobalToneMap::isValid function.
This will be called when the metadata comes over IPC, to reject any
metadata that is not spec compliant. This will need to be fixed on the
Chromium side.

Bug: 497439141
Change-Id: I854b83b004523471d0e06c9bb4a05b8a1bd20fd7
Reviewed-on: https://skia-review.googlesource.com/c/skia/+/1207416
Commit-Queue: Christopher Cameron <ccameron@google.com>
Reviewed-by: Christopher Cameron <ccameron@google.com>
Reviewed-by: Kaylee Lubick <kjlubick@google.com>

diff --git a/include/private/SkHdrMetadata.h b/include/private/SkHdrMetadata.h
index d54d0e36ee63dccfd4690bf5dcbf9e3240c89480..8a731dfdd4816ef211375587492dd04e126f35ee 100644
--- a/include/private/SkHdrMetadata.h
+++ b/include/private/SkHdrMetadata.h
@@ -242,6 +242,11 @@ struct SK_API AdaptiveGlobalToneMap {
      */
     SkString toString() const;
 
+    /**
+     * Return true if this metadata satisfies all normative constraints.
+     */
+    bool isValid() const;
+
     bool operator==(const AdaptiveGlobalToneMap& other) const {
         return fHdrReferenceWhite == other.fHdrReferenceWhite &&
                fHeadroomAdaptiveToneMap == other.fHeadroomAdaptiveToneMap;
diff --git a/src/codec/SkHdrAgtm.cpp b/src/codec/SkHdrAgtm.cpp
index 76c0aa2554feb7dc275a416ee81558f01215f264..18fa7d5897908669fb576574f89475c50c07feb4 100644
--- a/src/codec/SkHdrAgtm.cpp
+++ b/src/codec/SkHdrAgtm.cpp
@@ -286,7 +286,7 @@ MakeGainCurveXYMImage(const AdaptiveGlobalToneMap::HeadroomAdaptiveToneMap& hatm
     // Write the X, Y, and M values of the control points into the colors of the rows.
     SkBitmap bm32;
     bm32.allocPixels(SkImageInfo::Make(
-            AdaptiveGlobalToneMap::GainCurve::kMaxNumControlPoints, hatm.fAlternateImages.size(),
+            maxNumControlPoints, hatm.fAlternateImages.size(),
             kRGBA_F32_SkColorType, kPremul_SkAlphaType));
     for (size_t a = 0; a < hatm.fAlternateImages.size(); ++a) {
         const auto& alt = hatm.fAlternateImages[a];
@@ -797,5 +797,9 @@ SkString AdaptiveGlobalToneMap::toString() const {
     return result;
 }
 
+bool AdaptiveGlobalToneMap::isValid() const {
+    return AgtmHelpers::Validate(*this);
+}
+
 }  // namespace skhdr
 
